src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+
+    private Object[] elements;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfFull();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        if (checkAddIndex(index)) {
+            growIfFull();
+            elements = addResize(index);
+            elements[index] = value;
+            size++;
+        }
+    }
 
+    private Object[] addResize(int index) {
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        return elements;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
+        checkIndex(index);
+        elements[index] = value;
 
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public T remove(T element) {
+        for (int i = 0; i < size; i++) {
+            if (element == elements[i] || element != null && element.equals(elements[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("There is no such an element " + element);
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(int index) {
+        checkIndex(index);
+        T removeValue = (T) elements[index];
+        elements = removeResize(index);
+        elements[--size] = null;
+        return removeValue;
+    }
+
+    private Object[] removeResize(int index) {
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        return elements;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size <= 0;
+    }
+
+    private void growIfFull() {
+        if (isFull()) {
+            int newCapacity = size + (size >> 1);
+            Object[] newElements = new Object[newCapacity];
+            System.arraycopy(elements, 0, newElements, 0, size);
+            elements = newElements;
+        }
+    }
+
+    private boolean checkIndex(int index) {
+        if (index < 0 || index > size - 1) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds for length " + size);
+        }
+        return true;
```

Comment:
```suggestion
    private void checkIndex(int index) {
        if (index < 0 || index > size - 1) {
            throw new ArrayListIndexOutOfBoundsException("Index " + index
                    + " out of bounds for length " + size);
        }
```
you can make this method void instead have a lot of code in if block